### ISSUE TITLE
Bug: AST is `undefined` on Windows when more than one file with upper-case char in filename

### ISSUE THREAD

**Issue by nbolton** [typescript-eslint/typescript-eslint#8267]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

n/a

### Repro Code

The problem is a Windows filesystem issue, so can't repro in playground.

The code actually doesn't matter, as the issue is related to filenames in Windows.

### ESLint Config

```javascript
{
  "root": true,
  "overrides": [
    {
      "parserOptions": {
        "project": ["./tsconfig.json"]
      },
      "files": ["*.ts"],
      "extends": [
        "plugin:@typescript-eslint/recommended-requiring-type-checking"
      ]
    }
  ]
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "esModuleInterop": true,
    "strict": true
  }
}
```

### Expected Result

See attached .zip file.

[ts-sandbox.zip](https://github.com/typescript-eslint/typescript-eslint/files/13968822/ts-sandbox.zip)

1. Have filenames `Foo.ts` and `Bar.ts`
2. Both files should produce lint error: `Unsafe return of an `any` typed value  @typescript-eslint/no-unsafe-return`

### Actual Result

Only `Bar.ts` is linted, and `Foo.ts` incorrectly shows an error related to the file not being included in `tsconfig.json`.

```
> npx eslint .

C:\Projects\symless\ts-sandbox\src\Bar.ts
  2:3  error  Unsafe return of an `any` typed value  @typescript-eslint/no-unsafe-return

C:\Projects\symless\ts-sandbox\src\Foo.ts
  0:0  error  Parsing error: ESLint was configured to run on `<tsconfigRootDir>/src\Foo.ts` using `parserOptions.project`: <tsconfigRootDir>/tsconfig.json
However, that TSConfig does not include this file. Either:
- Change ESLint's list of included files to not include this file
- Change that TSConfig to include this file
- Create a new TSConfig that includes this file and include it in your parserOptions.project
See the typescript-eslint docs for more info: https://typescript-eslint.io/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file

✖ 2 problems (2 errors, 0 warnings)
```

### Additional Info

Searched issues for "upper case chars windows" and found no matches.

[ts-sandbox.zip](https://github.com/typescript-eslint/typescript-eslint/files/13968822/ts-sandbox.zip)

Workaround: Rename all files to lower case (bug does not occur with lower case).

I also rasied a bug with eslint (https://github.com/eslint/eslint/issues/18003), but they redirected me here.

Some pertinent details:

![image](https://github.com/typescript-eslint/typescript-eslint/assets/313427/c932a193-a2df-4690-aa41-3a95c7e26b21)

> This bug doesn't seem to be relevant to ESLint config or source file. It's a transient hard to reproduce issue related to the Windows filesystem. I have access to multiple Windows computers, and it only happens on one of them.

> I narrowed the problem down to `node_modules\@typescript-eslint\typescript-estree\dist\create-program\getWatchProgramsForProjects.js`:
```
        if (fileList.has(filePath)) {
            log('Found existing program for file. %s', filePath);
            updatedProgram ??= existingWatch.getProgram().getProgram();
            // sets parent pointers in source files
            updatedProgram.getTypeChecker();
            return [updatedProgram];
        }
```

> For some reason, `updatedProgram` doesn't contain the source file that is being linted:
```
updatedProgram.getSourceFiles().map(file=>file.fileName).join(", ")
```

> The filename is missing from this list.

**Comment by nbolton** [comment]
I'm unable to reproduce this on my Windows VM. It also doesn't happen on our Windows CI, and the rest of the team can't repro it on their Windows dev machines. 

Pretty sure this is just a weird problem related to my Windows install, so you can go ahead and close if you think it's not possible to reproduce.

**Comment by nbolton** [comment]
Oh... that's interesting. I moved it to a different hard drive (D:)

And now it works...

![image](https://github.com/typescript-eslint/typescript-eslint/assets/313427/f2ccdbe9-73bd-4a4d-b1aa-7bd6b68d8db4)

**Comment by bradzacher** [comment]
I cannot repro this either - your repro zip required some editing to get it to work - but otherwise it did not repro.

It's likely you're probably seeing something like #4680

**Comment by nbolton** [comment]
I think the hard drive must have been corrupt. `chkdsk` didn't show any errors, but when I was copying my projects from C to D drive, I got a ton of read errors.
